homeassistant/components/tplink_lte.py

```diff
@@ -0,0 +1,103 @@
+"""
+Support for TP-Link LTE modems.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/tplink_lte/
+"""
+import asyncio
+import logging
+from datetime import timedelta
+
+import aiohttp
+import attr
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_HOST, CONF_PASSWORD, EVENT_HOMEASSISTANT_STOP)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.aiohttp_client import async_create_clientsession
+
+REQUIREMENTS = ['tp-connected==0.0.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=10)
+
+DOMAIN = 'tplink_lte'
+DATA_KEY = 'tplink_lte'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+    })])
+}, extra=vol.ALLOW_EXTRA)
+
+
+@attr.s
+class ModemData:
+    """Class for modem state."""
+
+    modem = attr.ib()
+
+
+@attr.s
+class LTEData:
+    """Shared state."""
+
+    websession = attr.ib()
+    modem_data = attr.ib(init=False, factory=dict)
+
+    def get_modem_data(self, config):
+        """Get the requested or the only modem_data value."""
+        if CONF_HOST in config:
+            return self.modem_data.get(config[CONF_HOST])
+        if len(self.modem_data) == 1:
+            return next(iter(self.modem_data.values()))
+
+        return None
+
+
+async def async_setup(hass, config):
+    """Set up TP-Link LTE component."""
+    if DATA_KEY not in hass.data:
+        websession = async_create_clientsession(
+            hass, cookie_jar=aiohttp.CookieJar(unsafe=True))
+        hass.data[DATA_KEY] = LTEData(websession)
+
+    tasks = [_setup_lte(hass, conf) for conf in config.get(DOMAIN, [])]
+    if tasks:
+        await asyncio.wait(tasks)
+
+    return True
+
+
+async def _setup_lte(hass, lte_config, delay=0):
+    """Set up a TP-Link LTE modem."""
+    import tp_connected
+
+    if delay:
+        await asyncio.sleep(delay)
```

Comment:
We should probably cancel the `_setup_lte` task upon home assistant stop event and catch `CancelledError` here and around `modem.login`. Otherswise there will be a stacktrace when closing home assistant if we're still in this delay loop.

Reply:
So, if I understand it correctly:
```python
try:
    if delay:
        await asyncio.sleep(delay)
except asyncio.CancelledError:
    return
```

and

```python
try:
    await modem.login(password=password)
except asyncio.CancelledError:
    return
```

should do the thing, right?

Reply:
We also need to store the task that is created on failure so we can register a listener to home assistant stop event that can call a function that calls `task.cancel` if the task isn't done.

Reply:
The updated code I would write is:

```python
try:
    if delay:
        task = await asyncio.sleep(delay)
        await task
except asyncio.CancelledError:
    return
```

and

```python
try:
    task = await modem.login(password=password)
    await task
except asyncio.CancelledError:
    return
```

and cancel the task inside the `cleanup` function

Reply:
I don't think that works. Coroutones don't return tasks by default. We have to use a function from the asyncio library to wrap a coroutine in a task.
https://docs.python.org/3/library/asyncio-task.html#asyncio.Task

I would just save the task that is returned here:
https://github.com/home-assistant/home-assistant/pull/17111/files#diff-5410e9b3cb190ba123d66d6583931ad4R93

Reply:
Oh yes you're right about coroutines but I'm not sure about storing only the task from here: https://github.com/home-assistant/home-assistant/pull/17111/files#diff-5410e9b3cb190ba123d66d6583931ad4R93

Doesn't mean that we'll miss the first iteration for each modem setup? I'm talking about the tasks created here:
https://github.com/home-assistant/home-assistant/pull/17111/files#diff-5410e9b3cb190ba123d66d6583931ad4R68

Reply:
We won't miss that since those tasks are already tracked by the home assistant core. There is also no delay for those tasks. So unless `modem.login` takes a very long time to return or error, we should be fine.